src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -0,0 +1,159 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal abstract class JsonSeparatorNamingPolicy : JsonNamingPolicy
+    {
+        private readonly bool _lowercase;
+        private readonly char _separator;
+
+        internal JsonSeparatorNamingPolicy(bool lowercase, char separator) =>
+            (_lowercase, _separator) = (lowercase, separator);
+
+        public sealed override string ConvertName(string name)
+        {
+            // Rented buffer 20% longer that the input.
+            int rentedBufferLength = (12 * name.Length) / 10;
+            char[]? rentedBuffer = rentedBufferLength > JsonConstants.StackallocCharThreshold
+                ? ArrayPool<char>.Shared.Rent(rentedBufferLength)
+                : null;
+
+            int resultLength = 0;
+            Span<char> result = rentedBuffer is null
+                ? stackalloc char[JsonConstants.StackallocCharThreshold]
+                : rentedBuffer;
+
+            void ExpandBuffer(ref Span<char> result)
+            {
+                char[] newBuffer = ArrayPool<char>.Shared.Rent(result.Length * 2);
+
+                result.CopyTo(newBuffer);
+
+                if (rentedBuffer is not null)
+                {
+                    result.Slice(0, resultLength).Clear();
+                    ArrayPool<char>.Shared.Return(rentedBuffer);
+                }
+
+                rentedBuffer = newBuffer;
+                result = rentedBuffer;
+            }
+
+            void WriteWord(ReadOnlySpan<char> word, ref Span<char> result)
+            {
+                if (word.IsEmpty)
+                {
+                    return;
+                }
+
+                Span<char> destination = result.Slice(resultLength != 0
+                    ? resultLength + 1
```

Comment:
How do we know we have enough to do Slice + 1? slice will throw if it's too small. If it's checked elsewhere can you add assert here?